### ISSUE TITLE
eeRead may result in use-after-free on description buffers

### ISSUE THREAD

**Issue by pablosun** [snmishra/ftd2xx#64]
**Describe the bug**

The `eeRead()` method in `ftd2xx.py` may return a `ft_program_data` that contains addresses pointing to objects already being freed.

**To Reproduce**
On Python 3.9.13, when programming EEPROM on FT232RL devices, the following code might result in corrupted "Manufacturer" strings:

~~~ python
import ftd2xx as ftd
d = ftd.open(0)
data = d.eeRead()
data.Cbus0 = 0x20
d.eeProgram(data)
~~~

The resulting `Cbus0` field is correctly updated, but the string descriptors such as Manufacturer and Serial number may sometimes get corrupted after eeProgram().

**Expected behavior**

class FTD2xx probably need to store the reference to string buffers created by ctypes.create_string_buffer() to make sure these buffers are valid after returning `ft_program_data` to caller in `eeRead()`

**Desktop (please complete the following information):**
 - OS: Windows
 - OS Version: Windows 10
 - Python Version: 3.9.13
 - FTDI driver version: 135734 (reported from getDriverVersion()

**Additional context**
N/A

**Comment by snmishra** [comment]
@pablosun I think you're right. I don't have time to work on this now, but if you have a PR, I'd be happy to merge it.

**Comment by schlagmueller** [comment]
found that the issue was introduces with `1.2.0-beta.5` it worked with `1.2.0-beta.4`

The following code change _fixes_ it on my system - but I don't know really why and whether it really fixes the issue or just somehow works.

```
progdata = ft_program_data(
    **ProgramData(
        Signature1=0,
        Signature2=0xFFFFFFFF,
        Version=2,
        Manufacturer=c.cast(Manufacturer, c.c_char_p),
        ManufacturerId=c.cast(ManufacturerId, c.c_char_p),
        Description=c.cast(Description, c.c_char_p),
        SerialNumber=c.cast(SerialNumber, c.c_char_p),
    )
)
``` 

instead of

```
progdata = ft_program_data(
    **ProgramData(
        Signature1=0,
        Signature2=0xFFFFFFFF,
        Version=2,
        Manufacturer=c.addressof(Manufacturer),
        ManufacturerId=c.addressof(ManufacturerId),
        Description=c.addressof(Description),
        SerialNumber=c.addressof(SerialNumber),
    )
)
```

Any Ideas, thoughts, comments are welcome...

**Comment by schlagmueller** [comment]
Why the above works I don't know...

adding 
```
            progdata._Manufacturer = Manufacturer
            progdata._ManufacturerId = ManufacturerId
            progdata._Description = Description
            progdata._SerialNumber = SerialNumber
```
stores the references of the buffers within `progdata` and the garbage collector will not collect it early...
not pretty - but works

**Comment by pablosun** [comment]
I confirm that this fix works:

~~~
            progdata._Manufacturer = Manufacturer
            progdata._ManufacturerId = ManufacturerId
            progdata._Description = Description
            progdata._SerialNumber = SerialNumber
~~~

and the fix

~~~
c.cast(Manufacturer, c.c_char_p)
~~~

look like a better solution to me. It should work because 

~~~
addressof(some_ctypes_object)
~~~

returns an object of `<class 'int'>`, which does not keep the reference count to the `some_ctypes_object`.

`c.cast` returns an object of type `POINTER(type(obj))` and would actually keeps the reference count of the "pointed memory block", as stated in https://docs.python.org/3/library/ctypes.html#ctypes.cast.

So I think the `c.cast` would be a better solution.

**Comment by pablosun** [comment]
Hi @schlagmueller @snmishra a PR is created, thanks a lot for the help.

**Comment by snmishra** [comment]
Thanks @pablosun . Indeed the old version of the library used `ctypes.cast`.  Thanks for the PR

**Comment by snmishra** [comment]
Fixed in #67
